### PR TITLE
Sequential Container Module

### PR DESCRIPTION
**Pull request by pluskid**
Chain multiple modules.

Is there any easy way to tell whether a `Symbol` contains a loss function? Currently I am asking the user to explicitly specify `take_labels=True`.

BTW: @tqchen While running the example `sequential_module.py`, I run into the following segfaults. However, when I re-run it, I cannot reproduce it again. Do you have anything in mind what/where might be some potential bug here? Thanks!

```
[12:52:15] src/io/iter_mnist.cc:91: MNISTIter: load 60000 images, shuffle=1, shape=(100,784)
[12:52:16] src/io/iter_mnist.cc:91: MNISTIter: load 10000 images, shuffle=1, shape=(100,784)
INFO:root:Epoch[0] Train-accuracy=0.130850
INFO:root:Epoch[0] Time cost=1.362
INFO:root:Epoch[0] Validation-accuracy=0.225300
Process 6707 stopped
* thread #2: tid = 0x2b387, 0x00000001035a14de libmxnet.so`mshadow::Stream<mshadow::gpu>::Wait() + 30, stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x00000001035a14de libmxnet.so`mshadow::Stream<mshadow::gpu>::Wait() + 30
libmxnet.so`mshadow::Stream<mshadow::gpu>::Wait:
->  0x1035a14de <+30>: movq   (%rdi), %rdi
    0x1035a14e1 <+33>: callq  0x1038ae9e6               ; symbol stub for: cudaStreamSynchronize
    0x1035a14e6 <+38>: movl   %eax, %r15d
    0x1035a14e9 <+41>: testl  %r15d, %r15d
(lldb) bt
* thread #2: tid = 0x2b387, 0x00000001035a14de libmxnet.so`mshadow::Stream<mshadow::gpu>::Wait() + 30, stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x00000001035a14de libmxnet.so`mshadow::Stream<mshadow::gpu>::Wait() + 30
    frame #1: 0x0000000103745d40 libmxnet.so`mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::'lambda'(mxnet::RunContext, mxnet::engine::CallbackOnComplete)::operator()(mxnet::RunContext, mxnet::engine::CallbackOnComplete) const + 928
    frame #2: 0x000000010374577d libmxnet.so`std::__1::__function::__func<mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::'lambda'(mxnet::RunContext, mxnet::engine::CallbackOnComplete), std::__1::allocator<mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::'lambda'(mxnet::RunContext, mxnet::engine::CallbackOnComplete)>, void (mxnet::RunContext, mxnet::engine::CallbackOnComplete)>::operator()(mxnet::RunContext&&, mxnet::engine::CallbackOnComplete&&) + 45
    frame #3: 0x00000001035abce7 libmxnet.so`mxnet::engine::ThreadedEngine::ExecuteOprBlock(mxnet::RunContext, mxnet::engine::OprBlock*) + 279
    frame #4: 0x00000001035ae71e libmxnet.so`std::__1::__function::__func<mxnet::engine::ThreadedEnginePerDevice::PushToExecute(mxnet::engine::OprBlock*, bool)::'lambda'()::operator()() const::'lambda'(), std::__1::allocator<mxnet::engine::ThreadedEnginePerDevice::PushToExecute(mxnet::engine::OprBlock*, bool)::'lambda'()::operator()() const::'lambda'()>, void ()>::operator()() + 62
    frame #5: 0x00000001035ac1b8 libmxnet.so`void* std::__1::__thread_proxy<std::__1::tuple<std::__1::function<void ()> > >(void*) + 104
    frame #6: 0x00007fff96e5699d libsystem_pthread.dylib`_pthread_body + 131
    frame #7: 0x00007fff96e5691a libsystem_pthread.dylib`_pthread_start + 168
    frame #8: 0x00007fff96e54351 libsystem_pthread.dylib`thread_start + 13
```

### CODE REVIEW

**Comment by tqchen** [comment]
Seems to be a memory corruption. Could be some hidden BUG in the threading engine. I am not very sure. So far we haven't been able to get a quick reproducible code for the bad access bug. I will proof read the async engine code again

**Comment by mli** [comment]
a quick fix may be sleep 1s when existing to let the system have time to delete things

**Comment by pluskid** [comment]
@mli @tqchen Although not always reproducible, it happens quite frequently in this particularly example. Also it is in the middle of training, not when the system is shutting down. Just to double check, it is OK to have multiple executors running on the same context simultaneously? (They are not sharing memory or anything, just belong to different submodules).

**Comment by tqchen** [comment]
I think it is OK to have multiple executors to run on the same context. If this causes a problem, this should be seen as a bug. @pluskid can you make a simplified version of code that can reproduce the fault frequently?

**Comment by pluskid** [comment]
@tqchen I am not sure if running multiple executors is the reason. But let me prepare a simple script for this.

**Comment by pluskid** [comment]
@tqchen Here is a bit more detailed output from a backtrace with debug compilation. If you use [my branch](https://github.com/pluskid/mxnet/tree/container), the simple example [sequential_module.py](https://github.com/pluskid/mxnet/blob/container/example/module/sequential_module.py) trigger a seg-fault for many times on my machine. (I'm running on one single cpu context). Can you help debugging this?

```
* thread #2: tid = 0x37e9c, 0x00007fff9a3e2f06 libsystem_kernel.dylib`__pthread_kill + 10, stop reason = signal SIGABRT
  * frame #0: 0x00007fff9a3e2f06 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff96e594ec libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff9096e6e7 libsystem_c.dylib`abort + 129
    frame #3: 0x00007fff87ae1041 libsystem_malloc.dylib`free + 425
    frame #4: 0x000000010373c137 libmxnet.so`std::__1::__vector_base<mxnet::engine::ThreadedVar*, std::__1::allocator<mxnet::engine::ThreadedVar*> >::~__vector_base() [inlined] std::__1::__deallocate(__ptr=0x000000010036ece0) + 423 at new:164
    frame #5: 0x000000010373c12e libmxnet.so`std::__1::__vector_base<mxnet::engine::ThreadedVar*, std::__1::allocator<mxnet::engine::ThreadedVar*> >::~__vector_base() [inlined] std::__1::allocator<mxnet::engine::ThreadedVar*>::deallocate(this=0x000000010081dad8, __p=0x000000010036ece0, (null)=1) + 8 at memory:1664
    frame #6: 0x000000010373c126 libmxnet.so`std::__1::__vector_base<mxnet::engine::ThreadedVar*, std::__1::allocator<mxnet::engine::ThreadedVar*> >::~__vector_base() [inlined] std::__1::allocator_traits<std::__1::allocator<mxnet::engine::ThreadedVar*> >::deallocate(__a=0x000000010081dad8, __p=0x000000010036ece0, __n=1) + 24 at memory:1447
    frame #7: 0x000000010373c10e libmxnet.so`std::__1::__vector_base<mxnet::engine::ThreadedVar*, std::__1::allocator<mxnet::engine::ThreadedVar*> >::~__vector_base(this=0x000000010081dac8) + 382 at vector:472
    frame #8: 0x000000010373bf85 libmxnet.so`std::__1::vector<mxnet::engine::ThreadedVar*, std::__1::allocator<mxnet::engine::ThreadedVar*> >::~vector(this=0x000000010081dac8 size=1) + 21 at iterator:1244
    frame #9: 0x000000010373bf65 libmxnet.so`std::__1::vector<mxnet::engine::ThreadedVar*, std::__1::allocator<mxnet::engine::ThreadedVar*> >::~vector(this=0x000000010081dac8 size=1) + 21 at iterator:1244
    frame #10: 0x000000010373bf25 libmxnet.so`mxnet::engine::ThreadedOpr::~ThreadedOpr(this=0x000000010081da80) + 37 at threaded_engine.h:194
    frame #11: 0x000000010373bef5 libmxnet.so`mxnet::engine::ThreadedOpr::~ThreadedOpr(this=0x000000010081da80) + 21 at threaded_engine.h:194
    frame #12: 0x000000010373ae11 libmxnet.so`mxnet::common::ObjectPool<mxnet::engine::ThreadedOpr>::Delete(this=0x000000010036fcb0, ptr=0x000000010081da80) + 33 at object_pool.h:139
    frame #13: 0x00000001037386fd libmxnet.so`mxnet::common::ObjectPoolAllocatable<mxnet::engine::ThreadedOpr>::Delete(ptr=0x000000010081da80) + 29 at object_pool.h:196
    frame #14: 0x0000000103737d6d libmxnet.so`mxnet::engine::ThreadedEngine::OnComplete(this=0x000000010036fd70, threaded_opr=0x000000010081da80) + 1725 at threaded_engine.cc:362
    frame #15: 0x00000001037318b8 libmxnet.so`mxnet::engine::ThreadedEngine::OnCompleteStatic(engine=0x000000010036fd70, threaded_opr=0x000000010081da80) + 40 at threaded_engine.cc:368
    frame #16: 0x0000000103699790 libmxnet.so`mxnet::engine::CallbackOnComplete::operator(this=0x0000700000392740)() const + 32 at engine.h:41
    frame #17: 0x0000000103c84cd0 libmxnet.so`mxnet::GraphExecutor::GetOpExecEntry(this=0x0000000100375fb8, ctx=(stream = 0x0000000000000000), on_complete=CallbackOnComplete @ 0x0000700000392740)::'lambda'(mxnet::RunContext, mxnet::engine::CallbackOnComplete)::operator()(mxnet::RunContext, mxnet::engine::CallbackOnComplete) const + 3216 at graph_executor.cc:283
    frame #18: 0x0000000103c8402f libmxnet.so`void std::__1::__invoke_void_return_wrapper<void>::__call<mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::'lambda'(mxnet::RunContext, mxnet::engine::CallbackOnComplete)&, mxnet::RunContext, mxnet::engine::CallbackOnComplete>(mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::'lambda'(mxnet::RunContext, mxnet::engine::CallbackOnComplete)&&&, mxnet::RunContext&&, mxnet::engine::CallbackOnComplete&&) [inlined] decltype(__f=0x0000000100375fb8, __args=0x0000700000392878, __args=0x00007000003928a0)::'lambda'(mxnet::RunContext, mxnet::engine::CallbackOnComplete)&>(fp)(std::__1::forward<mxnet::RunContext, mxnet::engine::CallbackOnComplete>(fp0))) std::__1::__invoke<mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::'lambda'(mxnet::RunContext, mxnet::engine::CallbackOnComplete)&, mxnet::RunContext, mxnet::engine::CallbackOnComplete>(mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::'lambda'(mxnet::RunContext, mxnet::engine::CallbackOnComplete)&&&, mxnet::RunContext&&, mxnet::engine::CallbackOnComplete&&) + 175 at __functional_base:415
    frame #19: 0x0000000103c83fcb libmxnet.so`void std::__1::__invoke_void_return_wrapper<void>::__call<mxnet::GraphExecutor::GetOpExecEntry(__args=0x0000000100375fb8, __args=0x0000700000392878, __args=0x00007000003928a0)::'lambda'(mxnet::RunContext, mxnet::engine::CallbackOnComplete)&, mxnet::RunContext, mxnet::engine::CallbackOnComplete>(mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::'lambda'(mxnet::RunContext, mxnet::engine::CallbackOnComplete)&&&, mxnet::RunContext&&, mxnet::engine::CallbackOnComplete&&) + 75 at __functional_base:440
    frame #20: 0x0000000103c83cec libmxnet.so`std::__1::__function::__func<mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::'lambda'(mxnet::RunContext, mxnet::engine::CallbackOnComplete), std::__1::allocator<mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::'lambda'(mxnet::RunContext, mxnet::engine::CallbackOnComplete)>, void (mxnet::RunContext, mxnet::engine::CallbackOnComplete)>::operator(this=0x0000000100375fb0, __arg=0x0000700000392878, __arg=0x00007000003928a0)(mxnet::RunContext&&, mxnet::engine::CallbackOnComplete&&) + 76 at functional:1407
    frame #21: 0x000000010372a849 libmxnet.so`std::__1::function<void (mxnet::RunContext, mxnet::engine::CallbackOnComplete)>::operator(this=0x000000010081da80, __arg=(stream = 0x0000000000000000), __arg=CallbackOnComplete @ 0x00007000003928a0)(mxnet::RunContext, mxnet::engine::CallbackOnComplete) const + 185 at functional:1793
    frame #22: 0x000000010374880b libmxnet.so`mxnet::engine::ThreadedEngine::ExecuteOprBlock(this=0x000000010036fd70, run_ctx=(stream = 0x0000000000000000), opr_block=0x0000000100957500) + 411 at threaded_engine.h:290
    frame #23: 0x00000001037517ee libmxnet.so`void mxnet::engine::ThreadedEnginePerDevice::CPUWorker<(dmlc::ConcurrentQueueType)0>(this=0x000000010036fd70, block=0x0000000101ae1be0)0>*) + 94 at threaded_engine_perdevice.cc:180
    frame #24: 0x000000010375177f libmxnet.so`mxnet::engine::ThreadedEnginePerDevice::PushToExecute(this=0x0000000101a54378)::'lambda'()::operator()() const::'lambda'()::operator()() const + 31 at threaded_engine_perdevice.cc:76
    frame #25: 0x000000010375174d libmxnet.so`void std::__1::__invoke_void_return_wrapper<void>::__call<mxnet::engine::ThreadedEnginePerDevice::PushToExecute(mxnet::engine::OprBlock*, bool)::'lambda'()::operator()() const::'lambda'()&>(mxnet::engine::ThreadedEnginePerDevice::PushToExecute(mxnet::engine::OprBlock*, bool)::'lambda'()::operator()() const::'lambda'()&&&) [inlined] decltype(__f=0x0000000101a54378)::'lambda'()::operator()() const::'lambda'()&>(fp)(std::__1::forward<>(fp0))) std::__1::__invoke<mxnet::engine::ThreadedEnginePerDevice::PushToExecute(mxnet::engine::OprBlock*, bool)::'lambda'()::operator()() const::'lambda'()&>(mxnet::engine::ThreadedEnginePerDevice::PushToExecute(mxnet::engine::OprBlock*, bool)::'lambda'()::operator()() const::'lambda'()&&&) + 45 at __functional_base:415
    frame #26: 0x000000010375173c libmxnet.so`void std::__1::__invoke_void_return_wrapper<void>::__call<mxnet::engine::ThreadedEnginePerDevice::PushToExecute(__args=0x0000000101a54378)::'lambda'()::operator()() const::'lambda'()&>(mxnet::engine::ThreadedEnginePerDevice::PushToExecute(mxnet::engine::OprBlock*, bool)::'lambda'()::operator()() const::'lambda'()&&&) + 28 at __functional_base:440
    frame #27: 0x000000010375164c libmxnet.so`std::__1::__function::__func<mxnet::engine::ThreadedEnginePerDevice::PushToExecute(mxnet::engine::OprBlock*, bool)::'lambda'()::operator()() const::'lambda'(), std::__1::allocator<mxnet::engine::ThreadedEnginePerDevice::PushToExecute(mxnet::engine::OprBlock*, bool)::'lambda'()::operator()() const::'lambda'()>, void ()>::operator(this=0x0000000101a54370)() + 44 at functional:1407
    frame #28: 0x000000010374ac21 libmxnet.so`std::__1::function<void ()>::operator(this=0x0000000101a54370)() const + 129 at functional:1793
    frame #29: 0x000000010374a742 libmxnet.so`void* std::__1::__thread_proxy<std::__1::tuple<std::__1::function<void ()> > >(void*) [inlined] decltype(__f=0x0000000101a54370)> >(fp)(std::__1::forward<>(fp0))) std::__1::__invoke<std::__1::function<void ()> >(std::__1::function<void ()>&&) + 12 at __functional_base:415
    frame #30: 0x000000010374a736 libmxnet.so`void* std::__1::__thread_proxy<std::__1::tuple<std::__1::function<void ()> > >(void*) [inlined] void std::__1::__thread_execute<std::__1::function<void ()> >(__t=0x0000000101a54370, (null)=__tuple_indices<> @ 0x0000700000392e58)> >&, std::__1::__tuple_indices<>) + 22 at thread:337
    frame #31: 0x000000010374a720 libmxnet.so`void* std::__1::__thread_proxy<std::__1::tuple<std::__1::function<void ()> > >(__vp=0x0000000101a54370) + 368 at thread:347
    frame #32: 0x00007fff96e5699d libsystem_pthread.dylib`_pthread_body + 131
    frame #33: 0x00007fff96e5691a libsystem_pthread.dylib`_pthread_start + 168
    frame #34: 0x00007fff96e54351 libsystem_pthread.dylib`thread_start + 13
```

**Comment by piiswrong** [comment]
@pluskid after you are done with this refactoring and test it, could you change the tutorials to use modules and add a bucketing module tutorial?
We should also deprecate FeedForward some time.

**Comment by pluskid** [comment]
yes. i am thinking about how to organize the tutorial. looking to see how our new doc system looks like.

**Comment by piiswrong** [comment]
some comment I didn't have time to make before the previous PR is merged:
executor_group.py:13: _merge_multi_context. It's better to keep return type NDArray whether merge is on or not.
bucketing_module.py: switch_bucket need to borrow optimizer after creating new module for bucket

File: python/mxnet/module/sequential_module.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+        self.params_initialized = False
+        self.optimizer_initialized = False
+
+        return self # for easier chaining
+
+    @property
+    def data_names(self):
+        """A list of names for data required by this module."""
+        if len(self._modules) > 0:
+            return self._modules[0].data_names
```
Comment (piiswrong) [inline python/mxnet/module/sequential_module.py]: consider collecting all none-chained inputs for multibranch chaining?

File: python/mxnet/module/sequential_module.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+        return []
+
+    @property
+    def output_names(self):
+        """A list of names for the outputs of this module."""
+        if len(self._modules) > 0:
+            return self._modules[-1].output_names
```
Comment (piiswrong) [inline python/mxnet/module/sequential_module.py]: same with above

File: python/mxnet/module/sequential_module.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+    @property
+    def data_shapes(self):
+        """Get data shapes.
+        Returns
+        -------
+        A list of `(name, shape)` pairs. The data shapes of the
+        first module is the data shape of a `SequentialModule`.
+        """
+        assert self.binded
+        return self._modules[0].data_shapes
```
Comment (piiswrong) [inline python/mxnet/module/sequential_module.py]: same with above

**Comment by piiswrong** [comment]
We should support nonlinear (multi branch) chaining (in this PR or in later PRs). It shouldn't be very hard. Just keep a dictionary of free outputs and try to feed them to later modules with same named inputs.

**Comment by piiswrong** [comment]
https://github.com/dmlc/mxnet/issues/1846
the module sub folder needs to be added to setup.py
Currently it's not correctly installed.

**Comment by pluskid** [comment]
@piiswrong 
- `_merge_multi_context`: yes, I changed the return type to be always `NDArray` in order for the `SequentialModule` to work easily. Do you know if we have a `concatenate` operator for `NDArray`? Then we do not need to block async execution by numpy copying.
- switch bucket: yes, we have this logic (maybe hidden too deep) in `Module.bind`. If a the shared module provided for binding already installed an optimizer, `Module.bind` will automatically borrow it.

**Comment by pluskid** [comment]
@piiswrong Thanks for the suggestion about non-linear chaining. I was thinking about it. I tried to add a `rewiring` dict parameter for each module to be added that allow complicated re-mapping of input-output between two chained modules. But it soon become very complicated. I decided to keep it simple like this mainly because I want to avoid implementing a whole imperative computation graph system.

I think we are not advocating everything to be written in imperative way as what Torch does. Instead, main thing should still be in symbolic computation graph, which already had very non-linear "chaining". The imperative API is mainly used when one need complicated training logics, and then one can call those `forward`, `update` manually. With this philosophy, the primary goal of `SequentialModule` is actually as an example of how those intermediate level API could be used to operate multiple modules.

It just happens to be a `BaseModule` subclass (therefore high-level APIs like `fit` can be borrowed directly). It could do some interesting things like data-parallelism + model parallelism at the same time (as @mli mentioned the other day) -- I will add a simple example. But in general I think we might want to avoid it being abused.

**Comment by piiswrong** [comment]
@pluskid you don't need to asnumpy() to concatenate. You can allocate a new NDArray, slice along it's first dim and assign each of the outputs to a slice

File: example/module/sequential_module.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import logging
+
+# whether to demo model-parallelism + data parallelism
+demo_data_model_parallelism = True
+
+if demo_data_model_parallelism:
+    contexts = [[mx.context.gpu(0), mx.context.gpu(1)], [mx.context.gpu(2), mx.context.gpu(3)]]
+else:
+    contexts = [mx.context.cpu(), mx.context.cpu()]
+
```
Comment (pluskid) [inline example/module/sequential_module.py]: @mli @piiswrong here is a very simple hybrid model-par + data-par example. The whole network is split into two parts, the first part is computed in gpu0, gpu1, and the second part computed in gpu2, gpu3. We already have model-par in symbolic computation graph as in @tqchen 's lstm example. But I guess doing hybrid is a bit complicated in that API.

**Comment by pluskid** [comment]
@piiswrong Oh, yeah, you are absolutely right!

File: example/module/sequential_module.py (hunk 4 above)
Comment (mli) [inline example/module/sequential_module.py]: nice, thanks. i'm looking if we can get a machine translation example (embedding+lstm+attention) and test the model parallelism performance.

**Comment by tuonion** [comment]
Hi, @pluskid.  thanks for your work.

I am mxneter, I have a question about **model parallelism** : Can the mxnet split one layer into two gpus ? As far as i know, the system implement the model parallelism by binding one layer on one gpu.

**Comment by pluskid** [comment]
What do you mean splitting one layer to multiple GPUs? Do you actually mean data parallelism? Yes we could do data par with model par.

**Comment by tuonion** [comment]
@pluskid , it is not data-par . I  mean that : 

Assume that the system works on two gpus(gpu0,gpu1), we according to the following steps to create a network: 
1. Splitting  layer1 into two parts, one part binds on gpu0, and another part binds on gpu1; 
2. Splitting  layer2 into two parts, one part binds on gpu0, and another part binds on gpu1;
3. Doing the same  action with the rest of layers.

(As far as i know , the current mxnet implements the model-par by binding the whole layer1 on gpu1, and binding the whole layer2 on gpu2.)

**Comment by pluskid** [comment]
@tuonion What do you mean by splitting a layer into two parts? Do you mean, for example, for a fully connected layer, compute the first half of the output dimension in gpu0 and then the second half of the output dimension in gpu1?

**Comment by tuonion** [comment]
@pluskid . Yeah, that is what I want to express. I want to implement the model-par  in the mxnet like the Alex Krizhevsky implemented in [cuda-convnet2](https://github.com/akrizhevsky/cuda-convnet2). Is the current mxnet possible to implement it ? 

Thanks a lot.

**Comment by pluskid** [comment]
@tuonion I think you can just define **two** `FullyConnected` operator each with half of the output dimension and let them live on different GPUs. And then you can combine their results with `Concat`.

**Comment by tuonion** [comment]
@pluskid. Thank you for your good idea  very much, i will try this solution.

**Pull request merged**